### ISSUE TITLE
addNewlineTransformer escapes JSON

### ISSUE THREAD

**Issue by hscholz** [awslabs/lambda-streams-to-firehose#5]
My Kinesis stream already contains JSON records. These appear escaped in the s3 bucket if I use the standard newline transformer.

**Comment by IanMeyers** [comment]
Thanks for the note. In this case remove the JSON.stringify() call on line 104 and then rebuild and redeploy.

Ian

On 2 Mar 2016, at 13:36, Henning Scholz <notifications@github.com<mailto:notifications@github.com>> wrote:


My Kinesis stream already contains JSON records. These appear escaped in the s3 bucket if I use the standard newline transformer.

-
Reply to this email directly or view it on GitHub<https://github.com/awslabs/lambda-streams-to-firehose/issues/5>.



Amazon Web Services UK Limited. Registered in England and Wales with registration number 08650665 and which has its registered office at 60 Holborn Viaduct, London EC1A 2FD, United Kingdom.

**Comment by hscholz** [comment]
I did that already, it's just confusing and not so easy to use.

**Comment by rhamnett** [comment]
I think this should be made a configuration option, I have just come across this myself. The code was escaping the json and encapsulating each row in quotes.

If there was a config option kinesisDataType with option of "raw" or "json" or something similar this would be much better.

**Comment by IanMeyers** [comment]
Absolutely. I'll see if I can dig into that and make it easy to implement via environment settings.

**Comment by RobGoretsky** [comment]
@IanMeyers  - Thoughts on also making the Delivery Stream configurable by environment variable settings?  This would avoid the need to tag the Kinesis Stream with the destination Firehose Delivery Stream.  I have made this one tweak on lines 77-79 to allow me to pass in the target delivery stream as an Environment Variable named DELIVERY_STREAM:

```javascript
var deliveryStreamMapping = {
  'DEFAULT' : process.env.DELIVERY_STREAM
};

**Comment by rhamnett** [comment]
@RobGoretsky in our use case, I quite like the idea of being able to have multiple streams tagged and attached as multiple triggers to the single lambda function, so I am definitely in favour of the tagging. However as an override I suppose an ENV variable works too for a single stream.

**Comment by IanMeyers** [comment]
Hello,

@RobGoretsky agreed. I've just pushed a 1.5.0 deployable package that uses environment variables, which lets you set `StreamDatatype` and set values `CSV, CSV-WITH-NEWLINES, BINARY, JSON` as possible values and selects the correct transformer on the basis of the datatype. The implication of this is that you need deploy separate Lambda functions for each stream type being handled. This will be improved in a 1.5.x where we will attach the stream datatype to the kinesis stream that is configured as the Event Source.

Please test and let me know how it goes?

Thx,

Ian

**Comment by RobGoretsky** [comment]
@IanMeyers  - Thank you for the fast response!  It looks like this 1.5.0 version addresses @rhamnett's suggestion to make the StreamDatatype configurable via environment variable, but doesn't address my suggestion/request to make the Target Firehose Delivery Stream configurable via Environment variable.

@rhamnett  - I'm curious as to what advantage there is of running one Lambda function that consumes data from many Kinesis streams and emits to many Firehose Delivery streams, as opposed to having one Lambda function per source Kinesis stream, deployed with the same code.   There's no additional cost to having many Lambda functions with the same exact code, and splitting them out seems to allow for better monitoring/measurement in Cloudwatch..   Also, and this might be veering a bit off-topic here, so  happy to open a separate issue on this, but with this lambda-streams-to-firehose code in place, how do you search Cloudwatch Logs for errors?  I dont think all errors are prefixed with "ERROR" or anything easily searchable..

**Comment by rhamnett** [comment]
@IanMeyers nice work turning this around. The change actually hasn't solved the issue reported though as the point was that the jsonToStringTransformer should NOT wrap the data in JSON.stringify as this unnecessarily escapes it. Actually for directly saving json off the kinesis stream into s3 without transforming you just want to do exactly what the CSV type (addNewlineTransformer) does but it's confusing to be saving JSON with the CSV environment variable. Hope that makes sense! If you save JSON raw in S3 (non escaped) then you can for example have athena run over it by using the JSON serde. I don't see a reason why json would need to be JSON.stringified again.

@RobGoretsky you make a good point, my original thinking was just that...'thinking' :) I have not yet used multiple streams in a single function but your comments make me concur and we will probably use separate functions. Thanks.
